### PR TITLE
Add LPCWSTR support.

### PR DESCRIPTION
**Pull request by ikegami**
[Adding LCWSTR support to ::Lang::Win32](https://github.com/PerlFFI/FFI-Platypus/issues/291)

### CODE REVIEW

**Comment by ikegami** [comment]
I tested on Win and Linux. The above failures are due to problems with the script setting up the test, not the changes I made.

**Comment by plicease** [comment]
> I tested on Win and Linux. The above failures are due to problems with the script setting up the test, not the changes I made.

#293 should address the GH actions fail, due to a breaking change in GH actions recently.  The FreeBSD / cirrus CI fail seems legit, though I haven't investigated yet.

**Comment by plicease** [comment]
I've merged #293  can you please rebase on master when you have a chance to get that fix?

**Comment by ikegami** [comment]
`t/01_use.t` is giving `not ok 72 - test for FFI::Platypus::Type::Win32::LPCWSTR`
Not sure what that is. All the tests using the module actually work. 
I'm not getting on my systems. Then again, I didn't use zilla.
I gotta go for a while.

**Comment by ikegami** [comment]
oic. It looks like t/01_use.t is a generated file, and it's expects the test file to have a certain name. Will fix when I get back

**Comment by ikegami** [comment]
Yeah, a file named differently than expected was the only issue.

**Comment by plicease** [comment]
I am going to circle back to this sometime in the new year.  I've self assigned this PR for me to review so that I don't loose the thread.

File: lib/FFI/Platypus/Memory.pm [hunk 1]
```diff
@@ -103,9 +103,33 @@ useful in Perl.
...
+ my $size = strncpyW $opaque, $opaque, $max;
+
+The same as C<strncpy> above, except it works with 16-bit characters.
+This is useful when dealing with Windows Unicode strings (LPCWSTR and LPWSTR)
+and other UTF-16 libraries.
+
 =cut
 
-our @EXPORT = qw( malloc free calloc realloc memcpy memset strdup strndup strcpy );
+our @EXPORT = qw( malloc free calloc realloc memcpy memset strdup strndup strcpy strlenW strcpyW strncpyW );
```
Comment (plicease) [inline lib/FFI/Platypus/Memory.pm:132]: I'm conflicted on whether these should be `@EXPORT` or  `@EXPORT_OK`.

File: lib/FFI/Platypus/Memory.pm [hunk 2]
```diff
@@ -12,14 +12,14 @@ use base qw( Exporter );
 =head1 SYNOPSIS
 
  use FFI::Platypus::Memory;
- 
+
```
Comment (plicease) [inline lib/FFI/Platypus/Memory.pm:15]: I realize this is uncommon style, but this space is intentional because some (probably broken) POD readers render the literal blocks separately if you do not indent the space between.  I can fix this later though.

**Comment by plicease** [comment]
This is a fantastic PR.  I appreciate the attention to detail and the tests in particular.  Also I apologize for the long delay in this review, the last months have not left me with a lot of spare time to work on this.  I am onboard with all of the Win32 specific stuff.  I am not sure about the additions to the Memory module though.  The equivalent to `strlen` for wide strings appears to be `wcslen` and on Unix wchar_t is 32bit instead of 16 bit. I'd prefer the Memory module to be portable where possible, and to use the libc functions where available.

I'd like to get what works here merged, so maybe we can move the memory functions into `FFI::Platypus::Memory::Win32`?  Someone (possibly myself) can circle back later and take a stab at writing some portable wide string functions.

File: lib/FFI/Platypus/Memory.pm (hunk 1 above)
Comment (ikegami) [reply lib/FFI/Platypus/Memory.pm:132]: I never use `@EXPORT`, so I just followed suit. This becomes moot if they're moved to a separate module.

File: lib/FFI/Platypus/Memory.pm (hunk 2 above)
Comment (ikegami) [reply lib/FFI/Platypus/Memory.pm:15]: My editors automatically trim trailing spaces X_X

**Comment by ikegami** [comment]
> This is a fantastic PR. I appreciate the attention to detail and the tests in particular. Also I apologize for the long delay in this review, the last months have not left me with a lot of spare time to work on this. I am onboard with all of the Win32 specific stuff. I am not sure about the additions to the Memory module though. The equivalent to `strlen` for wide strings appears to be `wcslen` and on Unix wchar_t is 32bit instead of 16 bit. I'd prefer the Memory module to be portable where possible, and to use the libc functions where available.

Maybe different names would be better.

> I'd like to get what works here merged, so maybe we can move the memory functions into `FFI::Platypus::Memory::Win32`? Someone (possibly myself) can circle back later and take a stab at writing some portable wide string functions.

I purposefully avoided making them Windows-specified. Windows isn't the only user of UTF-16. JS uses UTF-16. C# uses UTF-16. These could definitely be split off, but I don't think ::Win32 would bee a good name.

**Comment by ikegami** [comment]
`strlenU16`, `strcpyU16`, `strncpyU16` and `::U16` or `::UTF16`?

File: lib/FFI/Platypus/Type/Win32/LPCWSTR.pm [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+sub perl_to_native_post
+{
+  pop @stack;
+}
+
+sub native_to_perl
+{
+   return unless defined $_[0];
+   return decode('UTF-16le', unpack('P'.(strlenW($_[0])*2), pack(_incantation, $_[0])));
```
Comment (plicease) [inline lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:77]: I think the libc function `wcslen`  or the windows native `lstrlenW` could be used here instead of strlenW.  Both return the number of characters if the string is in the BMP, and returns half the number of bytes if not.

It would mean that this code wouldn't be usable on non-windows but I don't think you would typically want to use it anywhere else.

**Comment by plicease** [comment]
The more I think about it, the more I'm not comfortable putting these functions into the Memory module as-is.  I see your point about ::Win32 (although I don't think C# or JS are easily callable from Platypus).  However, referring to these as UTF-16 functions does not sit well with me either, since for strings with non-BMP characters the "length" is in half the number of bytes.  I'd like to have whatever we come up with for dealing with wide character strings works naturally with non-Windows systems.  Hopefully without having different functions for each platform.

I would like to merge the LPCWSTR type plugin though so I propose we proceed with one of these solutions:

 1. Let's decouple the Memory changes from this PR entirely.  LPCWSTR plugin should be usable without them, so long as it internally uses... I'm thinking `lstrlenW` (or maybe `wcslen`) from system library.
 2. same as (1) except keep your `strlenW` implementation, but make it private.
 3. Memory module could expose the `wcslen` function, which the LPCWSTR plugin could use.  I'm not crazy about this function tbh, but the libc name (and implementation) for this function makes sense since the other functions in the Memory module are using the standard libc functions.  It's behavior on windows seems weird to me, but it is documented.

I think I am leaning for (2) because even though it means extra C code to bundle, the plugin is still usable on non-Windows platforms.  Although it won't normally be useful on non-Windows, platforms, it will make it easier to test the code from Unix.  For (1) and (3) the plugin should die if it is applied on a non-Windows platform.

Either way I think for now we remove `strcpyW` and `strncpyW`, again we can revisit wide string manipulation functions in another PR.  Or it might actually make more sense to put this in FFI::C, which doesn't yet have string manipulation classes, but my intent is to eventually add them there.

File: lib/FFI/Platypus/Type/Win32/LPCWSTR.pm [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+ use FFI::Platypus;
+
+ my $ffi = FFI::Platypus->new( api => 1 );
+ $ffi->load_custom_type('::Win32::LPCWSTR' => 'LPCWSTR');
+
+ $ffi->attach(takes_unicode_string => ['LPCWSTR'] => 'void');
+
+ my $str = "I ❤️ Platypus";
+ takes_unicode_string($str);
```
Comment (plicease) [inline lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:33]: Maybe we can use `MessageBoxExW` for the example here instead of including our own C code.

https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-messageboxexw

Comment (plicease) [reply lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:33]: or actually `MessageBoxW` is less complicated

https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-messageboxw

File: lib/FFI/Platypus/Lang/Win32.pm [hunk 5]
```diff
@@ -193,15 +193,45 @@ sub native_type_map
       }
     }
 
-    # stuff we are not yet dealing with
-    # LPCTSTR is unicode string, not currently supported
-    # LPWSTR 16 bit unicode string
-    # TBYTE TCHAR UNICODE_STRING WCHAR
+    # LPWSTR is simply mapped to opaque. Here are some recipies:
```
Comment (plicease) [inline lib/FFI/Platypus/Lang/Win32.pm:196]: Can you explain why we treat `LPWSTR` differently from `LPCWSTR`?  Is it normally a convention that LPWSTR is not null terminated?

**Comment by plicease** [comment]
The ::Lang::Win32 module should explicitly document `LPWSTR` and `LPCWSTR`.

**Comment by plicease** [comment]
> The ::Lang::Win32 module should explicitly document `LPWSTR` and `LPCWSTR`.

I'd also love to have a really good synopsis that actually calls some windows specific APIs.

File: lib/FFI/Platypus/Type/Win32/LPCWSTR.pm (hunk 3 above)
Comment (plicease) [reply lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:77]: See below: think I am leaning toward making your `strlenW` function private.

File: lib/FFI/Platypus/Lang/Win32.pm (hunk 5 above)
Comment (ikegami) [reply lib/FFI/Platypus/Lang/Win32.pm:196]: The "C" that's missing is for "constant". LWPSTR is a modifiable buffer. You provide space, the function fills it in.

File: lib/FFI/Platypus/Type/Win32/LPCWSTR.pm (hunk 4 above)
Comment (ikegami) [reply lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:33]: You want the test suite to popup a dialog and wait for the user to close it??? Smokers won't be able to run this, for one.

Comment (ikegami) [reply lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:33]: argh, this is just documentation. Sorry, just rushing a reply.

**Comment by ikegami** [comment]
> The ::Lang::Win32 module should explicitly document `LPWSTR` and `LPCWSTR`.

I kept the current style of documentation, which I frankly thought was useless. It doesn't tell the user what the module does/provides. All for changing this.

**Comment by ikegami** [comment]
> Either way I think for now we remove `strcpyW` and `strncpyW`

One of these -- or at least a public string length function so the user can implement it themselves  -- is needed to handle in-out string parameters. (See the "LPWSTR as in-out argument" test.) This is probably exceedingly rare, but do you know how much I'm frustrated with Win32::API for not providing some basic tools because they author probably didn't foresee their need?

File: lib/FFI/Platypus/Lang/Win32.pm (hunk 5 above)
Comment (ikegami) [inline lib/FFI/Platypus/Lang/Win32.pm:196]: "Recipes" is misspelled.

**Comment by plicease** [comment]
> I kept the current style of documentation, which I frankly thought was useless. It doesn't tell the user what the module does/provides. All for changing this.

I didn't think every variable type needs to be documented.  The Win32 API has frankly a stupid number of nested typedefs and it doesn't need to be repeated here.   This is new behavior for the wide string types.

> One of these -- or at least a public string length function so the user can implement it themselves -- is needed to handle in-out string parameters. (See the "LPWSTR as in-out argument" test.) This is probably exceedingly rare, but do you know how many times I've been m frustrated with Win32::API for not providing some basic tools because they author probably didn't foresee their need?

That is fair, but I think we can provide an interface which isn't Win32-centric.  I think FFI::C is the right place to do this.

File: lib/FFI/Platypus/Lang/Win32.pm [hunk 6]
```diff
@@ -39,9 +65,22 @@ sub abi
 
  my $hashref = FFI::Platypus::Lang::Win32->native_type_map;
 
-This returns a hash reference containing the native aliases for the
-Windows API.  That is the keys are native Windows API C types and the
-values are libffi native types.
+This is automatically called when C<< $ffi->lang('Win32'); >> is used.
+It returns definitions for the following native aliases for the Windows API,
+allowing them to be used in function declarations:
+C<BOOL>, C<BOOLEAN>, C<BYTE>, C<CCHAR>, C<CHAR>, C<COLORREF>, C<DWORD>, C<DWORDLONG>,
```
Comment (plicease) [inline lib/FFI/Platypus/Lang/Win32.pm:71]: I don't want all of these types explicitly listed in the documentation, as it is an extra thing to maintain, and may include dynamic parts in the future.  The old text described exactly what was returned, which I prefer, and if you want to include a note that explains how to get the list of available types you can use this one-liner:

`perl -MFFI::Platypus::Lang::Win32 -E "say for sort keys %{ FFI::Platypus::Lang::Win32->native_type_map }"`

File: lib/FFI/Platypus/Lang/Win32.pm [hunk 7]
```diff
@@ -26,6 +49,9 @@ C<uint32>.
 
  my $abi = FFI::Platypus::Lang::Win32->abi;
 
+This is automatically called when C<< $ffi->lang('Win32'); >> is used.
+It causes the proper ABI to make Windows system calls to be used.
```
Comment (plicease) [inline lib/FFI/Platypus/Lang/Win32.pm:53]:
```suggestion
It selects the appropriate ABI to make Win32 API function calls to be used.
```

File: lib/FFI/Platypus/Lang/Win32.pm [hunk 8]
```diff
@@ -193,15 +233,32 @@ sub native_type_map
       }
     }
 
-    # stuff we are not yet dealing with
-    # LPCTSTR is unicode string, not currently supported
-    # LPWSTR 16 bit unicode string
-    # TBYTE TCHAR UNICODE_STRING WCHAR
+    # We don't deal with WCHAR. It's simply a uint16_t, but it's
+    # semantically a Unicode Code Point. This is no biggie since one
+    # should never encounter a WCHAR outside of a LPWSTR or LPCWSTR.
```
Comment (plicease) [inline lib/FFI/Platypus/Lang/Win32.pm:238]: Lets make WCHAR an alias for `unsigned short`.  It may not be used often, but it doesn't hurt to expose it here.

File: lib/FFI/Platypus/Type/Win32/LPCWSTR.pm [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  ? 'Q'
+  : 'L!';
+
+my @stack;  # To keep buffer alive.
+
+sub perl_to_native
+{
+  if(defined $_[0])
+  {
+    my $buf = encode('UTF-16le', $_[0]."\0");
```
Comment (plicease) [inline lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:49]: Lets compute the wide string encoding instead of hard-coding it.  That way this will work for `wchar_t*` on UNIX as well as `LPCWSTR` on Windows.

I have some example code that can compute the encoding for most already supported platforms in practice, and we can back fill any others that we find not to work later.

https://gist.github.com/plicease/1e87ba3dd3faa8a0bd872b324672839d

File: lib/FFI/Platypus/Type/Win32/LPCWSTR.pm (hunk 3 above)
Comment (plicease) [reply lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:68]: Some here, compute the encoding instead of hard coding it.

File: lib/FFI/Platypus/Type/Win32/LPCWSTR.pm [hunk 10]
```diff
@@ -0,0 +1,123 @@
+package FFI::Platypus::Type::Win32::LPCWSTR;
```
Comment (plicease) [inline lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:1]:
```suggestion
package FFI::Platypus::Type::WideString;
```

Comment (plicease) [reply lib/FFI/Platypus/Type/Win32/LPCWSTR.pm:1]: Plus associated filename, etc changes.

File: t/ffi/lpcwstr.c [hunk 11]
```diff
@@ -0,0 +1,76 @@
+#include "libtest.h"
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+typedef uint16_t     WCHAR;
```
Comment (plicease) [inline t/ffi/lpcwstr.c:6]: I think we can just use the `wchar_t*` convention here instead of introducing the Windows aliases.  File can be called widestr.c

File: t/ffi/lpcwstr.c [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+strlenW(LPCWSTR s)
+{
+  size_t len = 0;
+  for (; *s; ++len, ++s);
+  return len;
+}
+
+
+EXTERN int
+lpcwstr_len(LPCWSTR s)
```
Comment (plicease) [inline t/ffi/lpcwstr.c:20]:
```suggestion
widestr_len(LPCWSTR s)
```

etc.

File: lib/FFI/Platypus/Memory.pm (hunk 1 above)
Comment (plicease) [inline lib/FFI/Platypus/Memory.pm:132]:
```suggestion
our @EXPORT = qw( malloc free calloc realloc memcpy memset strdup strndup strcpy wcslen wcscpy wcsncpy );
```

Comment (plicease) [reply lib/FFI/Platypus/Memory.pm:132]: We will keep it `@EXPORT` here, an explicit empty import is what I usually use anyway.

File: t/ffi/lpcwstr.c [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+typedef uint16_t     WCHAR;
+typedef WCHAR*       LPWSTR;
+typedef const WCHAR* LPCWSTR;
+
+static size_t
+strlenW(LPCWSTR s)
```
Comment (plicease) [inline t/ffi/lpcwstr.c:11]: Please use the system wcslen instead.

File: ffi/memory.c [hunk 14]
```diff
@@ -47,3 +52,33 @@ ffi_platypus_memory__strndup(const char *olds, size_t max)
   }
   return news;
 }
+
+EXPORT
+size_t
+ffi_platypus_memory__strlenW(LPCWSTR s)
```
Comment (plicease) [inline ffi/memory.c:58]: Please remove these we can use the system `wcs*` functions instead.

File: lib/FFI/Platypus/Memory.pm [hunk 15]
```diff
@@ -148,6 +172,10 @@ if($@)
   $ffi->attach([ ffi_platypus_memory__strndup => 'strndup' ] => ['string','size_t'] => 'opaque' => '$$');
 }
 
+$ffi->attach([ ffi_platypus_memory__strlenW => 'strlenW' ] => ['opaque'] => 'size_t' => '$');
+$ffi->attach([ ffi_platypus_memory__strcpyW => 'strcpyW' ] => ['opaque','opaque'] => 'opaque' => '$$');
+$ffi->attach([ ffi_platypus_memory__strncpyW => 'strncpyW' ] => ['opaque','opaque','size_t'] => 'opaque' => '$$$');
```
Comment (plicease) [inline lib/FFI/Platypus/Memory.pm:177]:
```suggestion
$ffi->attach( wcslen => ['opaque'] => 'size_t' => '$');
$ffi->attach( wcscpy => ['opaque','opaque'] => 'opaque' => '$$');
$ffi->attach( wcsncpy => ['opaque','opaque','size_t'] => 'opaque' => '$$$');
```

**Comment by plicease** [comment]
I will look into why the Perl 5.8.x fails are happening, I don't think it has to do with this PR.

The spelling errors can be fixed by adding the offending words to `author.yml`.

https://travis-ci.com/github/PerlFFI/FFI-Platypus/jobs/484209990#L306

**Comment by plicease** [comment]
>     3. Memory module could expose the `wcslen` function, which the LPCWSTR plugin could use.  I'm not crazy about this function tbh, but the libc name (and implementation) for this function makes sense since the other functions in the Memory module are using the standard libc functions.  It's behavior on windows seems weird to me, but it is documented.

So what I am thinking now is:

 4. Same as 3, but remove the windows specific stuff and use wchar_t* which works on Strawberry and Linux.  I think most already supported platforms should work.  On platforms that do not support `wcs*` and `wchar_t` we can disable this feature.

File: lib/FFI/Platypus/Lang/Win32.pm (hunk 8 above)
Comment (plicease) [reply lib/FFI/Platypus/Lang/Win32.pm:238]: This matches how `char` and `wchar_t` work.

**Comment by plicease** [comment]
I think it might be possible to write a type plugin that does the grunt work of an output string.  If not the type plugin API should maybe be extended so that we can.

**Comment by plicease** [comment]
> I think it might be possible to write a type plugin that does the grunt work of an output string. If not the type plugin API should maybe be extended so that we can.

POC for wide string output as a type plugin:
https://gist.github.com/plicease/1e87ba3dd3faa8a0bd872b324672839d#file-wide-string-out-pl

I think this would be easier to explain in the documentation than the nest of mallocs, frees and other function calls that you need to do manually and you no longer need `wcsn?cpy`.  I think there are situations where you might prefer to do that, but you still can via the libc API.

Because the type plugin API allows the type plugin to take arguments it could be just one plugin, and I think I prefer that probably.

File: lib/FFI/Platypus/Lang/Win32.pm [hunk 16]
```diff
@@ -11,8 +11,31 @@ use Config;
...
+
+ use constant MB_OK => 0;
+
+ my $ffi_user32 = FFI::Platypus->new( api => 1 );
+ $ffi_user32->lang('Win32');
+ $ffi_user32->find_lib( lib => 'user32' );
+
+ $ffi_user32->attach([MessageBoxW => 'MessageBox'] => ['HWND', 'LPCWSTR', 'LPCWSTR', 'UINT'] => 'int');
+
+ MessageBox(undef, "I ❤️ Platypus", "Confession", MB_OK);
```
Comment (plicease) [inline lib/FFI/Platypus/Lang/Win32.pm:25]: To get this to work I had to add `MB_DEFAULT_DESKTOP_ONLY` (not sure why) and `use utf8` (which makes sense tbh).

**Comment by plicease** [comment]
I realize that I made a lot of change requests here, and it is probably a little hard to follow because a number of these things are minor nits while others are substantial refactors, so I'm working on a branch based on your work, but with my suggested adjustments.  I think it is mostly functionally complete, but I want to fill in some of the documentation, and add a few more examples.  I welcome your feedback since it is based on your initial request and your work.  My implementation wouldn't have been possible without the work that you had already done.

#299 

The main thing that it does though is that it doesn't introduce any new C code, or add any public functions to the `Memory.pm` module, which was my main concern.  The way you pass read-write buffer strings is a tad awkward tbh, but as I mentioned earlier I think it is a little easier to explain than the more manual function approach.

Indeed there may actually be cases where the manual allocation/free is warranted, but I think you can still do that by writing your own bindings to the libc functions.  I've found that the types for the string functions in `Memory.pm` aren't quite sensible for all applications or general use so I'm hesitant to add more string functions there.

My PR also works on Unix with the native `wchar_t*` types and libc functions, and while I realize wide strings are not common there, I think while we are working on wide strings we may as well have them work "everywhere" for whatever approximation that "everywhere" even makes sense.

**Comment by ikegami** [comment]
I can follow. The issue is that I have no time right now, and that might be the case for a while.

**Comment by ikegami** [comment]
hmmm, I took a quick look. It looks lie you're assuming sizeof wchar_t == sizeof WCHAR, but that's not always the case.

**Comment by ikegami** [comment]
hmmm, I took a quick look. It looks like you're assuming sizeof wchar_t == sizeof WCHAR, but that's not always the case. A WCHAR is always 16 bits, while wchar_t can be as little as 8 bits, but also larger than 16 bits.

**Comment by ikegami** [comment]
Simple solution: Add an assert to ::Lang::Win32::load_custom_types

**Comment by plicease** [comment]
> hmmm, I took a quick look. It looks like you're assuming sizeof wchar_t == sizeof WCHAR, but that's not always the case. A WCHAR is always 16 bits, while wchar_t can be as little as 8 bits, but also larger than 16 bits.

`wchar_t` could be an arbitrary size equal or larger than `char` in general, but on Windows it is a 16bit wide UTF-16LE the same as `WCHAR`.

"The wchar_t type is an implementation-defined wide character type. In the Microsoft compiler, it represents a 16-bit wide character used to store Unicode encoded as UTF-16LE, the native character type on Windows operating systems"
https://docs.microsoft.com/en-us/cpp/cpp/char-wchar-t-char16-t-char32-t?view=msvc-160

`WCHAR` is in fact defined in terms of `wchar_t` if it is defined:
```c
#if !defined(_NATIVE_WCHAR_T_DEFINED)
typedef unsigned short WCHAR;
#else
typedef wchar_t WCHAR;
#endif
```
https://docs.microsoft.com/en-us/windows/win32/extensible-storage-engine/wchar

These statements also hold true for the gcc compiler used by Strawberry Perl and MSYS2.  

> Simple solution: Add an assert to ::Lang::Win32::load_custom_types

Sorry, I am not sure to what this is referring or what you would assert.

**Comment by ikegami** [comment]
> `WCHAR` is in fact defined in terms of `wchar_t` if it is defined:

Objection withdrawn. Misremembered.

> Sorry, I am not sure to what this is referring or what you would assert.

That sizeof wchar_t == sizeof WCHAR. But that's always the case.

**Comment by plicease** [comment]
I've merged #299 which is based on this PR and release Platypus as development version `1.35_01`.   Since that is a dev version we more wiggle room to make breaking changes until I cut a production release.  I went ahead and cut a dev release now so that I can start to accumulate reports from cpantesters.
